File: pkg/android/manifest_parser.go
package android

import (
	"encoding/xml"
	"os"
)

type Manifest struct {
	XMLName     xml.Name    `xml:"manifest"`
	Package     string      `xml:"package,attr"`
	Application Application `xml:"application"`
}

type Application struct {
	Name string `xml:"name,attr"` // `android:name` => use a custom struct to capture full namespace
}

func extractApplicationClassName(manifestPath string) (string, error) {
	data, err := os.ReadFile(manifestPath)
	if err != nil {
		return "", err
	}

	type AppWithAnyAttr struct {
		Attrs []xml.Attr `xml:",any,attr"`
	}

	type ManifestWrapper struct {
		XMLName     xml.Name       `xml:"manifest"`
		Package     string         `xml:"package,attr"`
		Application AppWithAnyAttr `xml:"application"`
	}

	var m ManifestWrapper
	if err := xml.Unmarshal(data, &m); err != nil {
		return "", err
	}

	var appName string
	for _, attr := range m.Application.Attrs {
		if attr.Name.Local == "name" {
			appName = attr.Value
			break
		}
	}

	return appName, nil
}

File: src/packages/android/manifest-parser.ts
import { readFile } from 'fs/promises';
import { parseStringPromise } from 'xml2js';

interface ManifestApplication {
  $?: {
    'android:name'?: string;
    [key: string]: any;
  };
}

interface Manifest {
  manifest?: {
    $?: {
      package?: string;
    };
    application?: ManifestApplication[];
  };
}

/**
 * Extracts the application class name from AndroidManifest.xml
 * Returns the value of android:name attribute from the <application> tag
 */
export async function extractApplicationClassName(manifestPath: string): Promise<string> {
  try {
    const data = await readFile(manifestPath, 'utf-8');
    const result: Manifest = await parseStringPromise(data);

    if (result.manifest?.application && result.manifest.application.length > 0) {
      const app = result.manifest.application[0];
      const appName = app.$?.[' android:name'] || app.$?.['android:name'] || '';
      return appName;
    }

    return '';
  } catch {
    return '';
  }
}
